File: solutions/cpp/0023.cpp
class Solution {
 public:
  ListNode* mergeKLists(vector<ListNode*>& lists) {
    ListNode dummy(0);
    ListNode* curr = &dummy;
    auto compare = [](const ListNode* a, const ListNode* b) {
      return a->val > b->val;
    };
    priority_queue<ListNode*, vector<ListNode*>, decltype(compare)> pq(compare);

    for (ListNode* list : lists)
      if (list)
        pq.push(list);

    while (!pq.empty()) {
      curr->next = pq.top(), pq.pop();
      curr = curr->next;
      if (curr->next) pq.push(curr->next);
    }

    return dummy.next;
  }
};

File: solutions/java/0023.java
class Solution {
  public ListNode mergeKLists(ListNode[] lists) {
    ListNode dummy = new ListNode(0);
    ListNode curr = dummy;
    Queue<ListNode> pq = new PriorityQueue<>((l1, l2) -> l1.val - l2.val);

    for (ListNode list : lists)
      if (list != null)
        pq.add(list);

    while (!pq.isEmpty()) {
      curr.next = pq.poll();
      curr = curr.next;
      if (curr.next != null)
        pq.add(curr.next);
    }

    return dummy.next;
  }
}